docs/codeql/codeql-language-guides/analyzing-data-flow-in-ruby.rst

```diff
@@ -0,0 +1,387 @@
+The global taint tracking library is used by extending the class ``TaintTracking::Configuration``:
+
+.. code-block:: ql
+
+   import codeql.ruby.DataFlow
+   import codeql.ruby.TaintTracking
+
+   class MyTaintTrackingConfiguration extends TaintTracking::Configuration {
+     MyTaintTrackingConfiguration() { this = "..." }
+
+     override predicate isSource(DataFlow::Node source) {
+       ...
+     }
+
+     override predicate isSink(DataFlow::Node sink) {
+       ...
+     }
+   }
+
+These predicates are defined in the configuration:
+
+-  ``isSource`` - defines where taint may flow from.
+-  ``isSink`` - defines where taint may flow to.
+-  ``isSanitizer`` - optionally, restricts the taint flow.
+-  ``isAdditionalTaintStep`` - optionally, adds additional taint steps.
+
+Similar to global data flow, the characteristic predicate (``MyTaintTrackingConfiguration()``) defines the unique name of the configuration and the taint analysis is performed using the predicate ``hasFlow(DataFlow::Node source, DataFlow::Node sink)``.
+
+Predefined sources and sinks
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+The data flow library contains a number of predefined sources and sinks, providing a good starting point for defining data flow based security queries.
+
+-  The class ``RemoteFlowSource`` (defined in module ``codeql.ruby.dataflow.RemoteFlowSources``) represents data flow from remote network inputs. This is useful for finding security problems in networked services.
+-  The library ``Concepts`` (defined in module ``codeql.ruby.Concepts``) contains several subclasses of ``DataFlow::Node`` that are security relevant, such as ``FileSystemAccess`` and ``SqlExecution``.
+
+For global flow, it is also useful to restrict sources to instances of ``LocalSourceNode``.
+The predefined sources generally do that.
+
+Class hierarchy
+~~~~~~~~~~~~~~~
+
+-  ``DataFlow::Configuration`` - base class for custom global data flow analysis.
+-  ``DataFlow::Node`` - an element behaving as a data-flow node.
+    -  ``DataFlow::LocalSourceNode`` - a local origin of data, as a data-flow node.
+    -  ``DataFlow::ExprNode`` - an expression behaving as a data-flow node.
+    -  ``DataFlow::ParameterNode`` - a parameter data-flow node representing the value of a parameter at method/block entry.
+
+    -  ``RemoteFlowSource`` - data flow from network/remote input.
+    -  ``Concepts::SystemCommandExecution`` - a data-flow node that executes an operating system command, for instance by spawning a new process.
+    -  ``Concepts::FileSystemAccess`` - a data-flow node that performs a file system access, including reading and writing data, creating and deleting files and folders, checking and updating permissions, and so on.
+    -  ``Concepts::Path::PathNormalization`` - a data-flow node that performs path normalization. This is often needed in order to safely access paths.
+    -  ``Concepts::CodeExecution`` - a data-flow node that dynamically executes Python code.
```

Comment:
Typo?

```suggestion
    -  ``Concepts::CodeExecution`` - a data-flow node that dynamically executes Ruby code.
```

Reply:
Good catch. This whole document was adapted from the equivalent doc for Python, so I just missed this.